### ISSUE TITLE
Heap memory usage increase during development builds after upgrading to v4.24.5

### ISSUE THREAD

**Issue by skatingincentralpark** [gatsbyjs/gatsby#36899]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

We have a large app in a monorepo using Gatsby **v4.15.2**.  Upon upgrading to **v4.24.5**, and running `gatsby develop` causes the build to crash during **Building development bundle** with a **Javascript heap out of memory** error. See error snapshot below:

#### Error Snapshot - Gatsby **v4.24.5** development build
```
<--- Last few GCs --->

[24620:0x160008000]    62793 ms: Mark-sweep (reduce) 3984.9 (4062.5) -> 3984.8 (4063.5) MB, 123.5 / 0.0 ms  (average mu = 0.191, current mu = 0.000) last resort GC in old space requested
[24620:0x160008000]    62929 ms: Mark-sweep (reduce) 3984.8 (4062.5) -> 3984.7 (4063.5) MB, 135.6 / 0.0 ms  (average mu = 0.098, current mu = 0.000) last resort GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: 0x100980a10 node::Abort() [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 2: 0x100980b74 node::OnFatalError(char const*, char const*) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 3: 0x100a999c4 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 4: 0x100a99958 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 5: 0x100c21f8c v8::internal::Heap::CollectionBarrier::Wait() [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 6: 0x100c2a71c v8::internal::Heap::SetUp() [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 7: 0x100bf9bb4 v8::internal::FactoryBase<v8::internal::Factory>::NewRawTwoByteString(int, v8::internal::AllocationType) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 8: 0x100e42e8c v8::internal::String::SlowFlatten(v8::internal::Isolate*, v8::internal::Handle<v8::internal::ConsString>, v8::internal::AllocationType) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
 9: 0x100aa7138 v8::internal::String::Flatten(v8::internal::Isolate*, v8::internal::Handle<v8::internal::String>, v8::internal::AllocationType) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
10: 0x100ab6c1c v8::String::Utf8Length(v8::Isolate*) const [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
11: 0x100961f58 node::Buffer::(anonymous namespace)::ByteLengthUtf8(v8::FunctionCallbackInfo<v8::Value> const&) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
12: 0x100b030d8 v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
13: 0x100b026d0 v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
14: 0x100b01f54 v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
15: 0x10122346c Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit [/Users/charleszhao/.nvm/versions/node/v14.17.6/bin/node]
16: 0x105b73080 
17: 0x105b7a928 
```

### Comparison between **'v4.15.2'**, and **'v4.24.5'**  dev build memory consumption  
For this test, we did the following:
1) Increased “heap size” allocated for memory by adding `NODE_OPTIONS=--max-old-space-size=8192` to our dev script
2) Installed process-top` to profile the memory usage

Our Gatsby **'v4.24.5'** build was now successful due to the increase in allocated heap memory from step **(2)** above. Also, we noticed an increase of ~2GB when comparing heap memory usage for **Gatsby v4.15.2**, and **Gatsby v4.24.5**, see table below:

| Gatsby Version | Heap Usage      |
|----------------|-----------------|
| Gatsby v4.15.2 | 2.5 GB / 2.6 GB |
| Gatsby v4.24.5 | 4.5 GB / 4.7 GB |

### Reproduction Link

https://github.com/skatingincentralpark/gatsby-v4-test

### Steps to Reproduce

1. git clone https://github.com/skatingincentralpark/gatsby-v4-test.git
2. cd to the project folder `/gatsby-v4-test`
3. Run `git switch 4.15.2-no-packages` to switch to the branch with Gatsby **v4.15.2**
4. Run `yarn` to install dependencies
5. Run `yarn dev` to start development server
6. Notice the highest heap will be around 623 MB / 657 MB
8. Run `yarn clean` to clear Gatsby cache
9. Run `git switch 4.24.5-no-packages` to switch to the branch with Gatsby **v4.24.5**
10. Run `yarn` to install dependencies
11. Run `yarn dev` to start development server
12. Notice the highest heap will be around 1.5 GB / 1.6 GB

### Expected Result

Expected the heap usage to be similar after upgrading to **v4.24.5** from **v4.15.2**.

### Actual Result

Heap usage increased by 1GB. This increased memory usage in Gatbsy **v4.24** causes a **'Javascript heap out of memory'** error if `NODE_OPTIONS=--max-old-space-size=8192` is not added to our dev script.

- **v4.15.2**: heap: 623 MB / 657 MB
```
cpu: 139.4% | rss: 898 MB (5.2%) | heap: 623 MB / 657 MB (94.9%) | ext: 336 MB | delay: 882 ms | 00:00:16 | 
```

- **v4.24.5**: heap: 1.5 GB / 1.6 GB
```
cpu: 130.1% | rss: 1.6 GB (9.4%) | heap: 1.5 GB / 1.6 GB (97.7%) | ext: 488 MB | delay: 891 ms | 00:00:27 | 
```

### Environment

```shell
System:
    OS: macOS 11.6
    CPU: (8) arm64 Apple M1
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.6 - ~/.nvm/versions/node/v14.17.6/bin/node
    Yarn: 1.22.18 - ~/.nvm/versions/node/v14.17.6/bin/yarn
    npm: 6.14.15 - ~/.nvm/versions/node/v14.17.6/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 106.0.5249.119
    Firefox: 104.0.2
    Safari: 15.0
  npmGlobalPackages:
    gatsby-cli: 4.10.1
```

**Comment by LekoArts** [comment]
Hi, thanks for the issue!

Also thanks for the detailed issue. I suppose you updated your app from 4.15 to 4.24 without any incremental updates right?
Because the version range is too large to be able to find the culprit. What would need to be done is bisect the releases, so e.g.

- Base: 4.15
- Update to 4.16, test
- Update to 4.17, test
- etc.

If then e.g. 4.17 would have it, double check that 4.16 is ok. Then one could be sure that the 4.17 release has this regression. If you have the time, we'd much appreciate it if you (or anyone else) could help with this.

**Comment by skatingincentralpark** [comment]
Hey, @LekoArts here are the results from the investigation (credits: @odunet ⭐️)!  We've also recorded the PartyTown Context error versions, but that can be ignored as the latest versions have implemented the fix.

### Gatsby versions investigation results **'v4.15.0** - **'v4.24.4'**
| Gatsby Version | Webpack 'pack' folder size | OOM Error? (Out Of Memory) | PartyTown Error? | Comment                                                        |
|----------------|----------------------------|----------------------------|-----------------|----------------------------------------------------------------|
| Gatsby v4.15.4 | 10.05GB                    | NO                         | NO              | Version usable with no errors                                  |
| Gatsby v4.16.0 | 10.05GB                    | NO                         | YES             | Version NOT usable with PartyTown Context error                |
| Gatsby v4.17.0 | 10.05GB                    | NO                         | YES             | Version NOT usable with PartyTown Context error                |
| Gatsby v4.17.2 | 10.05GB                    | NO                         | YES             | Version NOT usable with PartyTown Context error                |
| Gatsby v4.18.0 | 10.03GB                    | NO                         | YES             | Version NOT usable with PartyTown Context error                |
| Gatsby v4.18.2 | 10.05GB                    | NO                         | YES             | Version NOT usable with PartyTown Context error                |
| Gatsby v4.19.0 | 19.69GB                    | YES                        | YES             | Version NOT usable with PartyTown Context error, and OOM error |
| Gatsby v4.20.0 | 19.69GB                    | YES                        | NO              | version has OOM error, but usable with NODE_OPTIONS memory fix |
| Gatsby v4.21.0 | 19.69GB                    | YES                        | NO              | version has OOM error, but usable with NODE_OPTIONS memory fix |
| Gatsby v4.22.0 | 19.45GB                    | YES                        | NO              | version has OOM error, but usable with NODE_OPTIONS memory fix |
| Gatsby v4.23.0 | 19.69GB                    | YES                        | NO              | version has OOM error, but usable with NODE_OPTIONS memory fix |
| Gatsby v4.24.0 | 19.69GB                    | YES                        | NO              | version has OOM error, but usable with NODE_OPTIONS memory fix |

As shown above, Gatsby **v4.16.0** in commit https://github.com/gatsbyjs/gatsby/commit/a88703f4de47c0ba9db48914bc2e0df73440dc92 introduced a PartyTown Context bug, this bug rendered versions **v4.16.0** - **v4.19.0** unusable for us. This was later fixed in Gatsby **v4.20.0** in commit https://github.com/gatsbyjs/gatsby/commit/49cf094380bcf69d9239f8abbbd4db9c1968dcf8.

It can also be deduced from above that the OOM (Out of memory) error was introduced in Gatsby **v4.19.0**, the OOM error was accompanied by an increase of approximately **9GB** in the webpack 'pack' file size. 

Thank you ✊

**Comment by LekoArts** [comment]
Awesome, thank you for help on this! We'll have a closer look at this next week 👍

**Comment by dacevedo12** [comment]
Any updates on this? Seems like gatsby 5 is also affected

**Comment by LekoArts** [comment]
I've re-ran your reproduction again and I don't see the behavior described anymore. I think I saw it when I initially tried it but that could have been just a coincidance.

4.15.2:

![image](https://user-images.githubusercontent.com/16143594/201617580-192d3aa9-7c5a-4255-9910-64ce3dde5851.png)

4.24.5

![image](https://user-images.githubusercontent.com/16143594/201617642-bf261eff-2c75-46db-926e-f41f254ae1b2.png)

As you can see the heap is really similar. And this also makes sense. Node.js will use all the available memory and only then garbage collect.

We then tried to force garbage collection and also only saw 350 MB vs 400MB. You can try that for yourself:

```js
// in your gatsby-node.js

// ...rest of imports

const v8 = require(`v8`)
const vm = require(`vm`)
v8.setFlagsFromString(`--expose_gc`)
const gc = vm.runInNewContext(`gc`)

exports.createPages = async ({ graphql, actions, reporter }) => {
  setInterval(() => {
    // Prints out a string containing stats about your Node.js process.
    gc()
    console.log(top.toString())
  }, 5000)

  // ... rest of files
```

So the important question is: **Did you get out of memory errors?** Because if not then this behavior isn't too unusual.

**Comment by skatingincentralpark** [comment]
Hey @LekoArts, thank you very much for your response.  I just re-ran the reproduction (following the steps outlined in the issue) on two computers multiple times and am still seeing the behaviour described in the issue (which differs from your results).  

In answer to your question "**Did you get out of memory errors?**": Yes we do get OOM errors in our **main application** of the same structure (since it is much larger in size).  Which prevents us from starting our dev server.

- In our main application, adding `NODE_OPTIONS=--max-old-space-size=8192` allows the server to start, but memory usage is extremely high when developing, while shadowing files and [refreshing endpoints](https://www.gatsbyjs.com/docs/refreshing-content/), causing our computers to choke heavily. 
- It got so bad we downgraded back to **v4.15.2** until this is sorted. 
- Our worry is we'll be stuck on **v4.15.2** and unable to benefit from features in new releases which is commercially untenable as we will shortly have hundreds of sites live on Gatsby Cloud.

**I've created a reproduction of the OOM error in a new branch.  Instructions are at the bottom of the comment**

Your heap usage with forced garbage collection seems to align with my results.

### Questions
1. Could you please run the initial test again to confirm?  
3. Is forced garbage collection a proposed fix for this issue experienced?

## Summary of results
|              | Macbook Pro 13" | Macbook Pro 14" | Macbook Pro 13" - Force Garbage Collection |
|--------------|-----------------|-----------------|--------------------------------------------|
| v4.15.2 Heap | 605MB / 637MB   | 667MB / 710MB   | 318MB / 498MB                              |
| v4.24.5 Heap | 1.5GB / 1.6GB   | 1.3GB / 1.4GB   | 379MB / 559MB                              |

## Screenshots of results and System Info
### Macbook Pro 13"
![macbook pro 13 - 4 15 2](https://user-images.githubusercontent.com/61100546/202347146-21ec6897-205a-4184-826d-fedd3fdd4390.png)
![macbook pro 13 - 4 24 5](https://user-images.githubusercontent.com/61100546/201837214-917acd44-c6f8-44a6-a3a0-c8c24673c3fb.png)

```
System:
    OS: macOS 11.6
    CPU: (8) arm64 Apple M1
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.6 - ~/.nvm/versions/node/v14.17.6/bin/node
    Yarn: 1.22.18 - ~/.nvm/versions/node/v14.17.6/bin/yarn
    npm: 6.14.15 - ~/.nvm/versions/node/v14.17.6/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 106.0.5249.119
    Firefox: 104.0.2
    Safari: 15.0
  npmGlobalPackages:
    gatsby-cli: 4.10.1
```

### Macbook Pro 14"
<img width="938" alt="macbook pro 14 - 4 15 2" src="https://user-images.githubusercontent.com/61100546/202347288-00796a9b-c767-46ff-828c-c972efdbf3da.png">
<img width="944" alt="macbook pro 14 - 4 24 5" src="https://user-images.githubusercontent.com/61100546/201837224-ec11f400-5282-496a-bbcf-f32ea6363039.png">

```
  System:
    OS: macOS 12.1
    CPU: (10) arm64 Apple M1 Pro
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 18.6.0 - ~/.nvm/versions/node/v18.6.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.6.0/bin/yarn
    npm: 8.13.2 - ~/.nvm/versions/node/v18.6.0/bin/npm
  Languages:
    Python: 2.7.18 - /usr/bin/python
  Browsers:
    Chrome: 107.0.5304.110
    Safari: 15.2
  npmGlobalPackages:
    gatsby-cli: 4.24.0
```

### Macbook Pro 13" - Garbage Collected
<img width="1128" alt="4 15 2-force-garbage-collection-1" src="https://user-images.githubusercontent.com/61100546/202033435-97daf044-b527-4d7d-8308-ffdce0515cae.png">
<img width="1127" alt="4 24 5-force-garbage-collection-1" src="https://user-images.githubusercontent.com/61100546/202033453-e3ed54a5-5406-4dc3-a45b-4c33aeb1587c.png">

# Steps to produce OOM error
I have created a branch on the reproduction repo that will have the OOM error.  Below are the steps.

1. Run `git clone https://github.com/skatingincentralpark/gatsby-v4-test.git`
2. Cd to directory `cd gatsby-v4-test/`
1. Run `git switch 4.24.5-all-packages` to switch to the branch with Gatsby **v4.24.5** and various packages installed and initialised
5. Run `yarn` to install dependencies
6. Run `yarn dev` to start development server
7. OOM should appear

**Comment by pepijn-vanvlaanderen** [comment]
We started running into OOM issues as well, I will also post some of the testing we did. 

@skatingincentralpark maybe a workaround for you or something to test as well is with the DEV_SSR flag enabled. For us it made it possible again to start the dev server.

**Comment by skatingincentralpark** [comment]
@pepijn-vanvlaanderen 

Awesome, `DEV_SSR` does seem to work.  Although, it would be hard for us to use, since it seems to make using [shadowing](https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/shadowing/) files quite unpredictable 🥲.

Handy to have in the back pocket though.

**Comment by mjBayati** [comment]
the heap memory issue remained in gatsby5, 
in my experience with gatsby5, mdx2 and newest versions of gatsby_* packages, the report is as below

the scope:
1.less than 1000 mdx pages
2.using sharp and react svg for images

the raised problems:
1. development server with 12 gb of ram always raises heap allocation error with less than 1000 mdx pages
2. build takes at least 20 gb of ram
3. after serve, the images in pages take too much time to load in comparison with gatsby 4.15, at least 5 second!
4. the pages js bundle raised too much, that makes website to load after 5 second, in comparison with gatsby 4.15 that was taking less than 500 milisecond for same code.

I hope this comment be helpful, for those who wants to upgrade to reconsider their strategy,
Also I hope that development community revise the building process of gatsby5, webpack and react 18 together

**Comment by FraserThompson** [comment]
I have this issue too after updating from Gatsby 4.14.1 to Gatsby 5 (and to react-plugin-mdx). My site is pretty big, with around 1,000 pages and 17,000 images. None of the workarounds work and I can't build my site.

I have tried adjusting the heap limit to 16gb with `NODE_OPTIONS=--max-old-space-size=16384` (I have 32GB of system memory). `gatsby develop` works if I set the `DEV_SSR` flag, but `gatsby build` always runs out of memory at the `Building HTML renderer` phase with:

`FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`

**Comment by macsmel** [comment]
Any updates?

**Comment by mjBayati** [comment]
Is there anything that community(we) can help with?,
 by upgrading, the issues with `gatsby-plugin-mdx` is solved.
But now, we encounter heap memory issues and rendering issues while using `gatsby-plugin-material-ui`/

**Comment by macsmel** [comment]
> Is there anything that community(we) can help with?,
> by upgrading, the issues with `gatsby-plugin-mdx` is solved.
> But now, we encounter heap memory issues and rendering issues while using `gatsby-plugin-material-ui`/

I don't use gatsby-plugin-mdx and gatsby-plugin-material-ui in my projects, but I still have heap memory issues.

**Comment by mjBayati** [comment]
Hi, I commented before on this thread [here](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1319837951) and [here](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1364504895).
Now that i'm commenting, i can say that our issues with gatsby 5 are resolved.

the steps that we followed is as below:

1. there was recompilation per mdx, which when tracing the bug we found out that it's important to put your `import x.css` or all css related stuffs at the end of your imports.

2. it's so important to not concurrent build your template pages, for example your mdx's, because each of theme will be passed as context variable to template, which makes memory leak when you have accountable number of files to render.
3. if you are using some styling system like `gatsby-plugin-material-ui`, which raises warning at compile time, it's so important to care about that warnings
                     - for example in our case, using `gatsby-plugin-material-ui` added some hydration error while using dev_server, which with replacing it with it's newer verified version => `gatsby-plugin-emotion`, all errors has gone.
  
4. there is an important fix in `gatsby v5.3`, which is related to `esm` based modules i think, if you aren't yet, please upgrade to latest.
5. and the last and most important thing, please take a look at your gatsby-node, if you have different categories of files that should be compiled, separate them. and make their graphql query to run independently, and then create their related pages in some isolated context, it reduces your build time.
                        - for example, if you have blog and knowledge-base, where can be build in different templates, separate their build scenarios

**Comment by LekoArts** [comment]
We haven't tried the updated reproduction since my last comment, but already back then we couldn't verify what you were seeing unfortunately. The issue is also now a mix between "just" increased memory usage and OOM. It's not clear to us what the actual problem is now.

Anyways, as per https://www.gatsbyjs.com/docs/reference/release-notes/gatsby-version-support/ I'm afraid that we can't invest a lot of time chasing this down in the `4.x` release. For Gatsby 5 we did a lot of upgrades here and there, including https://github.com/gatsbyjs/gatsby/pull/37377 which updates `webpack` itself from a version from Oct 29, 2021 to the latest version.

So my asks are:

- Please only run the reproductions with `gatsby@next`
- Understand that this seems to be a "works on my machine"/"doesn't work on my machine" problem since we don't hear about this as a wide-spread problem
- It's time consuming to verify this and any additional help is appreciated

Thanks!

**Comment by sibelius** [comment]
this is happening to us on gatsby 5.4.2

**Comment by LekoArts** [comment]
> this is happening to us on gatsby 5.4.2

@sibelius Please understand that these types of comments are not helpful for us as they are not actionable. As mentioned in https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1385544620 we weren't able to reproduce this issue yet. Also "this" is quite ambiguous, what do you mean with "this"?

We want to help y'all but we need a https://www.gatsbyjs.com/contributing/how-to-make-a-reproducible-test-case/

**Comment by macsmel** [comment]
> we got this "fixed" like this
> 
> ```
> NODE_OPTIONS=--max_old_space_size=4096 gatsby build
> ```
> 
> giving 4GB for gatsby build

Sorry, but it is not helpful. The build shouldn't consume 2.5gb-3gb in one version and 4-10gb in new versions.

**Comment by pixelsoup** [comment]
Note: This response comes from the team that created this issue.  
@skatingincentralpark @McLeodSean @McLeodSean @pixelsoup 
cc @LekoArts 

---

After lots of testing we were able to fix our issue and successfully upgrade to **Gatsby v5.4.2** with the following changes.

### Early Tests
- Tests with Gatsby **v5.3.2** were promising but we would still get the  `FATAL ERROR: ... JavaScript heap out of memory` error when a site build exceeded **1,000 pages**
- The heap memory error always occurred during the `Building development bundle` process. The heap memory would slowly climb then _pop_ when memory exceeded 4.2GB - ish.
- adding `NODE_OPTIONS=--max-old-space-size=8192` to increase the upper limit of node memory would work OK with small sites (~100 pages), but it **didn't help** when we had a large number of pages (700+ or so). Increasing `max-old-space-size' is not an ideal fix anyway, but we would've taken it if it worked. 🤷‍♂️

### Conclusion
- Upgrading to Gatsby **v5.4.2** + **disabling sourceMaps** for  our emotion plugin `gatsby-plugin-emotion` works 
(turning off source maps is mentioned [here](https://www.gatsbyjs.com/docs/how-to/performance/resolving-out-of-memory-issues/#turn-off-source-maps-generation)).
- No memory issues occur even when when building a site with **1,800+ pages**.
- We suspect a combination of changes (probably webpack) in Gatsby 5 and the nature of our application (lots of shadowed css.js files for theming - ([see test repo](https://github.com/skatingincentralpark/gatsby-v4-test) related to this issue)  our emotion plugin `gatsby-plugin-emotion` was pushing the memory heap in local dev over 5GB causing the`JavaScript heap out of memory` error.
- Turning off sourceMaps stopped the memory heap failures. (we don't use them anyway 🤦🏼‍♂️).

### BEFORE 
- Source maps enabled. 
- `NODE_OPTIONS=--max-old-space-size=8192`

The JavaScript heap out of memory` error would trigger when memory exceeded 5GB. This happens for us with a site over **1,000 pages**. This is what we'd see just before the error on a large site. 

<img width="1317" alt="Screen Shot 2023-01-25 at 2 05 39 pm" src="https://user-images.githubusercontent.com/4155194/214471294-6c729587-abee-41ef-be3c-7b9280b28061.png">


### AFTER 
- Source maps disabled

**gatsby-config.js**
```js
    {
      resolve: 'gatsby-plugin-emotion',
      options: {
        sourceMap: false,
      },
    },
```
Instead of the memory heap climbing above 4GB it garbage collects at 4GB and dip back to a reasonable range. 

<img width="1156" alt="Screen Shot 2023-01-25 at 11 14 34 am" src="https://user-images.githubusercontent.com/4155194/214450721-4e9801de-5eec-4057-b156-477902d41237.png">

**Comment by pieh** [comment]
@pixelsoup

Using your repo I was able to narrow things down to introduction of [`Head`](https://github.com/gatsbyjs/gatsby/pull/35980) as commit that did increase memory usage during webpack compilation. I will be diving into it more, as that was hefty change with lot of moving pieces, but now I at least have something concrete to work against.

Thanks!

**Comment by panzacoder** [comment]
> Hi, I commented before on this thread [here](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1319837951) and [here](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1364504895). Now that i'm commenting, i can say that our issues with gatsby 5 are resolved.
> 
> the steps that we followed is as below:
> 
> 1. there was recompilation per mdx, which when tracing the bug we found out that it's important to put your `import x.css` or all css related stuffs at the end of your imports.
> 2. it's so important to not concurrent build your template pages, for example your mdx's, because each of theme will be passed as context variable to template, which makes memory leak when you have accountable number of files to render.
> 3. if you are using some styling system like `gatsby-plugin-material-ui`, which raises warning at compile time, it's so important to care about that warnings
>    - for example in our case, using `gatsby-plugin-material-ui` added some hydration error while using dev_server, which with replacing it with it's newer verified version => `gatsby-plugin-emotion`, all errors has gone.
> 4. there is an important fix in `gatsby v5.3`, which is related to `esm` based modules i think, if you aren't yet, please upgrade to latest.
> 5. and the last and most important thing, please take a look at your gatsby-node, if you have different categories of files that should be compiled, separate them. and make their graphql query to run independently, and then create their related pages in some isolated context, it reduces your build time.
>    - for example, if you have blog and knowledge-base, where can be build in different templates, separate their build scenarios
> 
> important note on running development server: **Change your `gatsby-node` in the way that your mdx related files not be rendered in development mode, there is an error when running development server on mdx files, for our team, it increases the memory usage incrementally while changing files and recompiling. which makes memory heap error after a while**.

I'm interested in several of the things you did here to mitigate your issues. This is the most unique advice I've seen on build performance with Gatsby, but some of the things you did are a little hard to determine how to replicate.

1. You say you moved you css import, can you be more specific? What file is it in? Where was it importing from? Did the css change or did you just change the import order?
2. How are you determining which templates are building concurrently? Did you go from sync to async node functions or did you determine some other form of control over concurrency?
5. You say "separate" the queries. How do you mean?  Are you writing multiple queries that need to be completed synchronously? I would love to see an example or reproduction, even if your example only contains example queries with no actual data, I'm interested in how your are structuring your file to achieve separation of concerns

Thanks!

**Comment by nkuehn** [comment]
@pixelsoup  thank you very much for the detailed reports and your conclusion.  Your conclusions (disabling all source map generation, granting more memory and some other optimizations) led us to a 50-60% decrease now but I would not really call that a "fix" because the memory requirements are still significantly higher than before, it's just bringing it back into a state that works at all. 

We are using gatsby 5.5, MDX v2 and emotion and are still struggling with peak memory issues in the *build* step.  The issue likely came up with one of the 4.x releases in our case and got amplified by upgrading MDX to v2 now, causing this investigation. It's small sites in terms of number of pages (< 200) but a nontrivial gatsby usage.

Specifically, the issue is happening in the "Building HTML renderer" step of the build and it has the sneaky property of the RAM only spiking up for less than a second if run on a cached build. I'm forcing GC now, too before logging but it's clear that the memory report is not only uncollected thrash but actual memory assignments.  
@LekoArts you mentioned that you had not been able to reproduce on a second run above,  there may be a possibility that it was due to the sampling of `process-top` via `setInterval` not catching the issue at the right moment if it was a cached build? 

I'm aware that by title this discussion is about issues in the development server, but I found that analyzing the problem via the build process is more predictable and better to debug.  

Once we got our users productive again through a mix of throwing hardware at the problem and the above mentioned optimizations we'll try to get to a reproduction scenario on a current gatsby version. May take a bit though.

**Comment by pieh** [comment]
Hey folks, I located a reason for increased memory usage (at least in repro provided by @pixelsoup ) and put up draft PR ( Building development bundle ) to address it somewhat (i.e. if you don't use Head API, you should get back to pre 4.19 memory usage levels).

This allowed me to go from (heap max - `heap: 3.0 GB`):
```
cpu: 145.7% | rss: 669 MB (4.0%) | heap: 474 MB / 504 MB (94.1%) | ext: 142 MB | delay: 901 ms | 00:00:10 | loadavg: 1.56, 0.71, 0.44
cpu: 142.1% | rss: 732 MB (4.4%) | heap: 519 MB / 561 MB (92.4%) | ext: 144 MB | delay: 1080 ms | 00:00:11 | loadavg: 1.52, 0.71, 0.44
cpu: 163.8% | rss: 794 MB (4.8%) | heap: 588 MB / 621 MB (94.6%) | ext: 148 MB | delay: 1280 ms | 00:00:12 | loadavg: 1.52, 0.71, 0.44
cpu: 144.1% | rss: 1.9 GB (11.4%) | heap: 1.5 GB / 1.8 GB (85.5%) | ext: 148 MB | delay: 1015 ms | 00:00:19 | loadavg: 1.48, 0.72, 0.45
cpu: 157.9% | rss: 2.3 GB (14.1%) | heap: 1.6 GB / 1.9 GB (88.1%) | ext: 502 MB | delay: 885 ms | 00:00:20 | loadavg: 1.48, 0.72, 0.45
cpu: 165.8% | rss: 2.9 GB (17.5%) | heap: 1.9 GB / 2.2 GB (86.2%) | ext: 742 MB | delay: 879 ms | 00:00:21 | loadavg: 1.44, 0.72, 0.45
cpu: 166.4% | rss: 3.0 GB (18.3%) | heap: 2.0 GB / 2.1 GB (92.5%) | ext: 880 MB | delay: 827 ms | 00:00:22 | loadavg: 1.44, 0.72, 0.45
cpu: 228.2% | rss: 3.0 GB (18.3%) | heap: 2.0 GB / 2.1 GB (92.2%) | ext: 880 MB | delay: 403 ms | 00:00:23 | loadavg: 1.44, 0.72, 0.45
cpu: 210.2% | rss: 3.0 GB (18.3%) | heap: 2.0 GB / 2.1 GB (92.6%) | ext: 880 MB | delay: 237 ms | 00:00:24 | loadavg: 1.44, 0.72, 0.45
cpu: 198.1% | rss: 3.0 GB (18.3%) | heap: 2.0 GB / 2.1 GB (92.3%) | ext: 880 MB | delay: 69 ms | 00:00:25 | loadavg: 1.44, 0.72, 0.45
cpu: 184.1% | rss: 4.9 GB (29.4%) | heap: 2.2 GB / 2.3 GB (95.1%) | ext: 2.5 GB | delay: 4 ms | 00:00:36 | loadavg: 1.34, 0.74, 0.46
cpu: 176.7% | rss: 4.9 GB (29.5%) | heap: 2.2 GB / 2.3 GB (95.1%) | ext: 2.5 GB | delay: 717 ms | 00:00:37 | loadavg: 1.34, 0.74, 0.46
cpu: 169.2% | rss: 4.9 GB (29.4%) | heap: 2.2 GB / 2.3 GB (95.4%) | ext: 2.5 GB | delay: 722 ms | 00:00:38 | loadavg: 1.34, 0.74, 0.46
cpu: 162.2% | rss: 4.9 GB (29.6%) | heap: 2.3 GB / 2.4 GB (96.1%) | ext: 2.5 GB | delay: 782 ms | 00:00:39 | loadavg: 1.34, 0.74, 0.46
cpu: 116.9% | rss: 5.0 GB (30.2%) | heap: 2.4 GB / 2.4 GB (97.7%) | ext: 2.5 GB | delay: 2614 ms | 00:00:40 | loadavg: 1.34, 0.74, 0.46
cpu: 131.3% | rss: 5.1 GB (30.9%) | heap: 2.5 GB / 2.6 GB (97.8%) | ext: 2.5 GB | delay: 572 ms | 00:00:42 | loadavg: 1.39, 0.76, 0.47
cpu: 140.2% | rss: 5.2 GB (31.5%) | heap: 2.6 GB / 2.6 GB (97.8%) | ext: 2.5 GB | delay: 852 ms | 00:00:43 | loadavg: 1.39, 0.76, 0.47
cpu: 151.1% | rss: 5.3 GB (32.1%) | heap: 2.7 GB / 2.7 GB (97.9%) | ext: 2.5 GB | delay: 769 ms | 00:00:44 | loadavg: 1.39, 0.76, 0.47
cpu: 151.2% | rss: 5.4 GB (32.7%) | heap: 2.8 GB / 2.8 GB (97.9%) | ext: 2.6 GB | delay: 819 ms | 00:00:45 | loadavg: 1.39, 0.76, 0.47
cpu: 145.6% | rss: 5.6 GB (34.0%) | heap: 3.0 GB / 3.0 GB (98.1%) | ext: 2.6 GB | delay: 888 ms | 00:00:47 | loadavg: 1.36, 0.76, 0.47
cpu: 178.4% | rss: 3.6 GB (21.8%) | heap: 427 MB / 626 MB (68.3%) | ext: 1.5 GB | delay: 786 ms | 00:00:48 | loadavg: 1.36, 0.76, 0.47
⠀
You can now view website in the browser.
```

to (heap max - `heap: 1.5 GB`)

```
cpu: 130.2% | rss: 563 MB (3.4%) | heap: 370 MB / 406 MB (91.3%) | ext: 142 MB | delay: 1143 ms | 00:00:09 | loadavg: 2.13, 1.29, 0.83
cpu: 149.5% | rss: 624 MB (3.8%) | heap: 424 MB / 462 MB (91.8%) | ext: 145 MB | delay: 1310 ms | 00:00:10 | loadavg: 2.13, 1.29, 0.83
cpu: 167.9% | rss: 692 MB (4.2%) | heap: 495 MB / 526 MB (94.1%) | ext: 148 MB | delay: 1103 ms | 00:00:11 | loadavg: 2.13, 1.29, 0.83
cpu: 154.3% | rss: 1.2 GB (7.0%) | heap: 950 MB / 1.0 GB (93.7%) | ext: 145 MB | delay: 859 ms | 00:00:14 | loadavg: 1.95, 1.27, 0.82
cpu: 163.2% | rss: 1.8 GB (10.6%) | heap: 1.2 GB / 1.3 GB (89.3%) | ext: 441 MB | delay: 831 ms | 00:00:15 | loadavg: 1.95, 1.27, 0.82
cpu: 164.4% | rss: 1.7 GB (10.3%) | heap: 1.1 GB / 1.2 GB (90.8%) | ext: 515 MB | delay: 774 ms | 00:00:16 | loadavg: 2.04, 1.30, 0.83
cpu: 214.0% | rss: 1.7 GB (10.3%) | heap: 1.1 GB / 1.2 GB (90.4%) | ext: 515 MB | delay: 307 ms | 00:00:17 | loadavg: 2.04, 1.30, 0.83
cpu: 204.2% | rss: 1.7 GB (10.3%) | heap: 1.1 GB / 1.2 GB (90.6%) | ext: 515 MB | delay: 99 ms | 00:00:18 | loadavg: 2.04, 1.30, 0.83
cpu: 198.3% | rss: 1.7 GB (10.3%) | heap: 1.1 GB / 1.2 GB (91.2%) | ext: 515 MB | delay: 4 ms | 00:00:19 | loadavg: 2.04, 1.30, 0.83
cpu: 194.7% | rss: 1.7 GB (10.3%) | heap: 1.1 GB / 1.2 GB (91.8%) | ext: 515 MB | delay: 10 ms | 00:00:20 | loadavg: 2.04, 1.30, 0.83
cpu: 174.8% | rss: 2.7 GB (16.3%) | heap: 1.3 GB / 1.3 GB (94.8%) | ext: 1.4 GB | delay: 15 ms | 00:00:26 | loadavg: 1.96, 1.30, 0.83
cpu: 157.5% | rss: 2.7 GB (16.4%) | heap: 1.3 GB / 1.3 GB (95.9%) | ext: 1.4 GB | delay: 732 ms | 00:00:27 | loadavg: 1.96, 1.31, 0.84
cpu: 154.3% | rss: 2.8 GB (17.0%) | heap: 1.4 GB / 1.4 GB (96.4%) | ext: 1.4 GB | delay: 790 ms | 00:00:29 | loadavg: 1.96, 1.31, 0.84
cpu: 133.7% | rss: 2.9 GB (17.4%) | heap: 1.5 GB / 1.5 GB (97.2%) | ext: 1.4 GB | delay: 2000 ms | 00:00:30 | loadavg: 1.96, 1.31, 0.84
cpu: 167.1% | rss: 1.6 GB (9.4%) | heap: 349 MB / 608 MB (57.4%) | ext: 169 MB | delay: 744 ms | 00:00:31 | loadavg: 1.96, 1.31, 0.84
cpu: 167.7% | rss: 1.6 GB (9.4%) | heap: 486 MB / 613 MB (79.3%) | ext: 175 MB | delay: 800 ms | 00:00:32 | loadavg: 1.88, 1.30, 0.84
cpu: 156.3% | rss: 1.7 GB (10.1%) | heap: 681 MB / 727 MB (93.6%) | ext: 183 MB | delay: 791 ms | 00:00:35 | loadavg: 1.88, 1.30, 0.84
⠀
You can now view website in the browser.
```

(`Building development bundle` step also takes less time as a result)

That won't help at all with build problem (mentioned by @nkuehn ), but is at least a start at addressing these problems

**Comment by nkuehn** [comment]
@pieh  thank you for helping out!  We do have a huge memory issue in the development mode, too.  I just turned to debugging the build first because it was more reproducible.  

**Caveat**: I still cannot reliably that my problem is the same.  My current path of investigation is that at build time I am seeing very large page-specific JS bundles in `.cache/page-ssr/routes/` that repeat the same dependent libraries over and over again whereas I think that dependencies that are not per-page should normally live only once in `render-page.js` (may point towards an issue our application structure).  Roughly looking into a heap dump showed that all these page specific js bundles live in memory in the peak memory situations (webpack `CachedSource` objects), which feels bogus even if our implementation has no issues.  I have not been able to heap dump a dev server though (OOM, dumping requires another doubling of memory).   --> all done on gatsby 5.5.0 and latest MDX v2

**Comment by mjBayati** [comment]
> > Hi, I commented before on this thread [here](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1319837951) and [here](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1364504895). Now that i'm commenting, i can say that our issues with gatsby 5 are resolved.
> > the steps that we followed is as below:
> > 
> > 1. there was recompilation per mdx, which when tracing the bug we found out that it's important to put your `import x.css` or all css related stuffs at the end of your imports.
> > 2. it's so important to not concurrent build your template pages, for example your mdx's, because each of theme will be passed as context variable to template, which makes memory leak when you have accountable number of files to render.
> > 3. if you are using some styling system like `gatsby-plugin-material-ui`, which raises warning at compile time, it's so important to care about that warnings
> >    
> >    * for example in our case, using `gatsby-plugin-material-ui` added some hydration error while using dev_server, which with replacing it with it's newer verified version => `gatsby-plugin-emotion`, all errors has gone.
> > 4. there is an important fix in `gatsby v5.3`, which is related to `esm` based modules i think, if you aren't yet, please upgrade to latest.
> > 5. and the last and most important thing, please take a look at your gatsby-node, if you have different categories of files that should be compiled, separate them. and make their graphql query to run independently, and then create their related pages in some isolated context, it reduces your build time.
> >    
> >    * for example, if you have blog and knowledge-base, where can be build in different templates, separate their build scenarios
> > 
> > important note on running development server: **Change your `gatsby-node` in the way that your mdx related files not be rendered in development mode, there is an error when running development server on mdx files, for our team, it increases the memory usage incrementally while changing files and recompiling. which makes memory heap error after a while**.
> 
> @mjBayati I'm interested in several of the things you did here to mitigate your issues. This is the most unique advice I've seen on build performance with Gatsby, but some of the things you did are a little hard to determine how to replicate.
> 
> 1. You say you moved you css import, can you be more specific? What file is it in? Where was it importing from? Did the css change or did you just change the import order?
> 2. How are you determining which templates are building concurrently? Did you go from sync to async node functions or did you determine some other form of control over concurrency?
> 3. You say "separate" the queries. How do you mean?  Are you writing multiple queries that need to be completed synchronously? I would love to see an example or reproduction, even if your example only contains example queries with no actual data, I'm interested in how your are structuring your file to achieve separation of concerns
> 
> Thanks!

Hi, thanks for your deep diving into my comment,
here it is some more explanation about your questions: 

1. it was about reordering imports, and putting all css imports to last,
2. I realized that in new version of mdx rendering, all content will be passed to template as `children`, which this content, is huge and increases memory usage, our build had below structure before
```
Promise(fetchBlogsContentFromGraphql() && renderBlogs())
Promise(fetchKbContentsFromGraphql() && renderKbs())
Promsie(fetchLandingsFromGraphql() && renderLandings())

Promise.all().then(doOtherWorks())
```

and also `renderX()` functions was as below:

```

async loop on each page content:
          renderPageByTemplate()

```
what we did was removing concurrent builds and `await` on each section, and also making `renderPagesLoops` sync.

** Also it's important to separate your `graphql query + build`, like what we did, (separations of blogs, kbs and landings)

by this kind of works, we was able to make our dev and production servers runnable!.

**Comment by nkuehn** [comment]
An indication to whomever may be competent to investigate the issue better:  Development mode with the discussed heap issues (< 50 pages project) creates an 11GB sized `./.cache/webpack/stage-develop/0.pack` file in our project that has similar issues as discussed here. Unfortunately I have no means to introspect it reasonably and also am not able to create a heap dump due to the sheer amount of memory involved.

**Comment by pieh** [comment]
I released `gatsby@5.7.0-next.1` with some improvements (mentioned in https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1419097423) - I would appreciate feedback if this change helped at all. If it did help I will look to release this in stable v5 and backport it to v4 stable as well.

> An indication to whomever may be competent to investigate the issue better: Development mode with the discussed heap issues (< 50 pages project) creates an 11GB sized `./.cache/webpack/stage-develop/0.pack` file in our project that has similar issues as discussed here. Unfortunately I have no means to introspect it reasonably and also am not able to create a heap dump due to the sheer amount of memory involved.

This won't fully address this, but after few test runs I saw that using this snippet in `gatsby-node` file:
```js
exports.onCreateWebpackConfig = ({ actions, stage, getConfig }) => {
  if (stage === `develop` || stage === `develop-html`) {
    actions.setWebpackConfig({
      cache: {
        compression: "gzip",
      },
    })
  }
}
```

keeps memory usage lower than default with no compression at all (it also does decrease cache files size on disk, but I don't think that was main issue and rather just symptom). `brotli` is an another compression option which results in even smaller files in cache, but in my tests memory usage was on par with default (no compression).

You can also try go even further and disable webpack caching completely which decreases memory usage in my tests quite a bit (but makes dev server restarts longer, so trade-off, but might be waranted to use for now if dev server just crashes on OOMs):

```js
exports.onCreateWebpackConfig = ({ actions, stage, getConfig }) => {
  if (stage === `develop` || stage === `develop-html`) {
    actions.setWebpackConfig({
      cache: false,
    })
  }
}
```

**Comment by nkuehn** [comment]
@pieh thank you!  Here's my results in a test site with our custom theme and ca 150 MDX pages across two templates, tracked as recommended above with a forced garbage collection before the "process-top" call.  

 - `5.5.0`:  peaks at `rss: 2.6 GB (15.2%) | heap: 1.9 GB / 2.8 GB (67.8%) | ext: 12 GB`. Note the huge "ext" amount (buffers?). Interestingly the ext memory drops back to 5GB after navigating to a second page. 
 - `5.7.0-next.1`  peaks at `rss: 2.6 GB (15.3%) | heap: 1.9 GB / 2.5 GB (75.1%) | ext: 12 GB`.  A slight heap improvement but it does not significantly address my ext memory issue.
 - `5.7.0-next.1` with gzip compression in the webpack cache: `rss: 2.0 GB (11.5%) | heap: 1.8 GB / 2.1 GB (83.2%) | ext: 12 GB`. Unchanged, but I saw that the persisted file is compressed.  
 - `5.7.0-next.1` with no webpack caching at all `rss: 1.6 GB (9.3%) | heap: 1.2 GB / 1.5 GB (80.5%) | ext: 8.8 GB`. Here's some improvement but I still wonder what webpack is doing in those 9GB.  
 
In any case, it takes a very long time until the first page is loaded in the dev preview but from then on things are fast - time that gets longer with the number of pages but only in the bundle generation stage.

Since you pointed to the cache settings I've tried out various combinations of webpack cache configuration and found the following to at least freeing a lot of the memory again after navigating to a second page in the dev server (not on the first page view though and only if the `DEV_SSR` flag is on). 
```js
idleTimeout: 100, // defaults to 60000 millis
idleTimeoutAfterLargeChanges: 100, // defaults to 1000 millis
allowCollectingMemory: true, // defaults to false.
maxMemoryGenerations: 0, // see docs https://webpack.js.org/configuration/cache/#cachemaxmemorygenerations
```

TL:Dr:  Improvements but nothing that looks like it's addressing a root cause.

**Comment by pieh** [comment]
For develop specifically we hope https://github.com/gatsbyjs/gatsby/pull/37040 (once we are able to resume work on it) could help a lot for larger sites as webpack would just skip compilatation for a lot of modules that are not needed for what the user need to display a single page on dev server. I've been doing some memory snapshoting / allocation timeline stuff and I definitely saw a lot of memory being allocated specifically during serialization (part of webpack's caching pipeline).

I will continue to dig into the entire pipeline for MDX overall (not just develop), but I don't have anything concrete yet on it.

**Comment by nkuehn** [comment]
@pieh  thanks!  concerning the MDX build memory issue my current hack is to import the template components that are used in many pages into `gatsby-ssr.js` without doing anything sensible there with them.  That resolved the build time  memory issue as all re-used code goes into the central renderer bundle.

**Comment by randychilau** [comment]
Hi All,

Thanks to everyone who provided comments and experiences in troubleshooting the memory issues for both development and production builds. I wanted to share my experience to add to the discussion and perhaps clarify some comments from others.

[The project](https://github.com/layer5io/layer5) is using Gatsby v.4.25.4, and the desire was to move to gatsby-plugin-mdx v4 (and MDX v2) to resolve [app bundle bloat issue](https://www.gatsbyjs.com/docs/reference/release-notes/v4.21/#gatsby-plugin-mdx-v4). Full Gatsby Info in the dropdown.

<details>
  <summary>Gatsby Info</summary>

```
System:<br />
    OS: macOS 12.6.3
    CPU: (4) x64 Intel(R) Core(TM) i5-5350U CPU @ 1.80GHz
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 20.3.1 - ~/.nvm/versions/node/v20.3.1/bin/node
    npm: 9.6.7 - ~/.nvm/versions/node/v20.3.1/bin/npm
  Browsers:
    Chrome: 114.0.5735.198
    Safari: 16.3
  npmPackages:
    gatsby: ^4.20.0 => 4.25.4
    gatsby-awesome-pagination: ^0.3.8 => 0.3.8
    gatsby-background-image: ^1.6.0 => 1.6.0
    gatsby-cli: ^5.5.0 => 5.5.0
    gatsby-plugin-anchor-links: ^1.2.1 => 1.2.1
    gatsby-plugin-feed: ^4.25.0 => 4.25.0
    gatsby-plugin-image: ^2.22.0 => 2.22.0
    gatsby-plugin-loadable-components-ssr: ^4.3.2 => 4.3.2
    gatsby-plugin-manifest: ^4.25.0 => 4.25.0
    gatsby-plugin-mdx: ^3.20.0 => 4.4.0
    gatsby-plugin-meta-redirect: github:layer5labs/gatsby-plugin-meta-redirect => 1.1.1
    gatsby-plugin-preload-fonts: ^3.24.0 => 3.25.0
    gatsby-plugin-robots-txt: ^1.7.1 => 1.7.1
    gatsby-plugin-sharp: ^4.25.1 => 4.25.1
    gatsby-plugin-sitemap: ^5.25.0 => 5.25.0
    gatsby-plugin-styled-components: ^5.25.0 => 5.25.0
    gatsby-plugin-svgr: ^3.0.0-beta.0 => 3.0.0-beta.0
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.30 => 1.1.30
    gatsby-redirect-from: ^0.5.0 => 0.5.0
    gatsby-source-filesystem: ^4.25.0 => 4.25.0
    gatsby-transformer-sharp: ^4.25.0 => 4.25.0
  npmGlobalPackages:
    gatsby-cli: 5.11.0
```

>note: for those who didn't know (like me), you can get the above info  with the [`gatsby info`](https://www.gatsbyjs.com/docs/reference/gatsby-cli/#info) command when you install Gatsby CLI

</details>

---

After making all the [migration changes](https://www.gatsbyjs.com/plugins/gatsby-plugin-mdx/#migrating-from-v3-to-v4), the production builds kept resulting in OOM heap memory errors as described by [@FraserThompson](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1331605013), [@nkuehn](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1418200435), and others.

As many noted, I was able to successfully execute a production build, by adding memory to process with the command:
`NODE_OPTIONS=--max_old_space_size=4096 gatsby build`

This felt like an unsatisfactory requirement for a version migration on a small-medium sized site, perhaps not practical for an open source project/contributors, and did not seem to resolve the cause of the memory issue.

---

I used the following tools while investigating the OOM issue:
- [process-top](https://www.npmjs.com/package/process-top), note: I wasn't able to get it to work for the entire build by using the command `npx process-top ./node_modules/.bin/gatsby build` (it would stop working at some point). Instead I installed the package and then added it to my gatsby-node.js file (code in the dropdown)
<details>
  <summary>process-top code</summary>

```
const top = require("process-top")();

setInterval(function () {
 console.log(top.toString())
}, 1000);

/*
 if you want to only see the memory string when it spikes above a certain amount, I used the following in the setInterval function instead:

 let heapMem = top.memory().heapUsed / 1000000;
  if (heapMem > 1500) {
    console.log(top.toString());
  }
*/

/*
if you want to log all these memory lines to a file for reference, you can maybe do this:

var fs = require('fs');
var path = require('path');
const top = require("process-top")();

async function log(){
    fs.appendFile(path.join(__dirname, 'memory_log.txt'), top.toString() + '\n', (err) => {
        if (err) throw err;
        logText = '';
    }); 
}
setInterval(log, 1000);

*/
```

</details>

- `GATSBY_WEBPACK_LOGGING="build-javascript,build-html" gatsby build` command as [described here](https://github.com/gatsbyjs/gatsby/pull/33214). I did not understand the output very well, but it gave more info than using `verbose`.
- [gatsby-plugin-webpack-bundle-analyser-v2](https://www.gatsbyjs.com/plugins/gatsby-plugin-webpack-bundle-analyser-v2/) and [gatsby-plugin-perf-budgets](https://github.com/pieh/gatsby-plugin-perf-budgets) (note: perf-budgets did not work with the production build).

I did not try the tools [described here](https://relative-ci.com/documentation/guides/webpack-stats/gatsby).

---

**What worked for me regarding the Production Build** (so far?)

I tried a lot of the suggestions here, and for me, the only thing that worked was @nkuehn's[ comment](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1422616312), and I didn't understand what was needed to be imported into the `gatsby-ssr.js` file and how make it useless.

In the end, I imported everything that I was using as a [shortcode](https://github.com/layer5io/layer5/blob/42bd0bc38856254bb0de92b9459b54baab181eff/root-wrapper.js), as well as all the template files themselves, and then creating an array of all the imported components. The project is open source, so I believe you can see the [file itself here](https://github.com/layer5io/layer5/blob/42bd0bc38856254bb0de92b9459b54baab181eff/gatsby-ssr.js), if not, it is here in the dropdown.
<details>
  <summary>gatsby-ssr.js</summary>

```
import CTA_ImageOnly from "./src/components/Call-To-Actions/CTA_ImageOnly";
import CTA_FullWidth from "./src/components/Call-To-Actions/CTA_FullWidth";
import CTA_Bottom from "./src/components/Call-To-Actions/CTA_Bottom";
import FAQ from "./src/sections/General/Faq";
import { Link } from "gatsby";
import { BlogWrapper } from "./src/collections/blog/Blog.style.js";
import { CareerWrapper } from "./src/collections/careers/Career.style.js";
import { BookWrapper } from "./src/collections/service-mesh-books/Book.style.js";
import { NewsWrapper } from "./src/collections/news/News.style.js";
import { ProjectWrapper } from "./src/collections/projects/Project.style.js";
import { ResourcesWrapper } from "./src/collections/resources/Resources.style.js";
import { ProgramsWrapper } from "./src/collections/programs/Programs.style.js";
import { ChapterStyle } from "./src/components/Learn-Components/Chapters-Style/chapters.style.js";
import { Col } from "./src/reusecore/Layout";
import Button from "./src/reusecore/Button";
import Blockquote from "./src/reusecore/Blockquote";
import BlockquoteAlt from "./src/reusecore/Blockquote/Blockquote-alt-style";

import blogPostTemplate from "./src/templates/blog-single.js";
import blogCategoryListTemplate from "./src/templates/blog-category-list.js";
import blogTagListTemplate from "./src/templates/blog-tag-list.js";
import EventsTemplate from "./src/templates/events.js";
import EventTemplate from "./src/templates/event-single.js";
import NewsPostTemplate from "./src/templates/news-single.js";
import BookPostTemplate from "./src/templates/book-single.js";
import ProgramPostTemplate from "./src/templates/program-single.js";
import MultiProgramPostTemplate from "./src/templates/program-multiple.js";
import CareerPostTemplate from "./src/templates/career-single.js";
import MemberTemplate from "./src/templates/member-single.js";
import MemberBioTemplate from "./src/templates/executive-bio.js";
import WorkshopTemplate from "./src/templates/workshop-single.js";
import LabTemplate from "./src/templates/lab-single.js";
import resourcePostTemplate from "./src/templates/resource-single.js";
import integrationTemplate from "./src/templates/integrations.js";
import sectionTemplate from "./src/sections/Learn-Layer5/Section/index.js";
import chapterTemplate from "./src/templates/learn-chapter.js";
import courseOverviewTemplate from "./src/templates/course-overview.js";
import courseListTemplate from "./src/templates/courses-list.js";
//
// import from "./src/sections/Careers/Careers-Programs-grid/index.js""
// );

//workaround to resolve build heap OOM error due to mdxv2 upgrade (ref: https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1422616312), revisit with upgrade to gatsby v5 and gatsby-plugin-mdx if still applicable
// eslint-disable-next-line no-unused-vars
const MDXComponents = [
  Col,
  FAQ,
  Link,
  Button,
  Blockquote,
  BlockquoteAlt,
  BlogWrapper,
  BookWrapper,
  ChapterStyle,
  CareerWrapper,
  NewsWrapper,
  ProgramsWrapper,
  ProjectWrapper,
  ResourcesWrapper,
  CTA_ImageOnly,
  CTA_FullWidth,
  CTA_Bottom,
  blogPostTemplate,
  blogCategoryListTemplate,
  blogTagListTemplate,
  EventsTemplate,
  EventTemplate,
  NewsPostTemplate,
  BookPostTemplate,
  ProgramPostTemplate,
  MultiProgramPostTemplate,
  CareerPostTemplate,
  MemberTemplate,
  MemberBioTemplate,
  WorkshopTemplate,
  LabTemplate,
  resourcePostTemplate,
  integrationTemplate,
  sectionTemplate,
  chapterTemplate,
  courseOverviewTemplate,
  courseListTemplate
];

export { onRenderBody } from "./onRenderBody";
export { wrapRootElement } from "./root-wrapper";
export { wrapPageElement } from "./page-wrapper";
```

</details>

---

After rejoicing about the production build working, I ran into issues with the development builds where it would successfully build (but take a long time) and if I started to browse pages:
- the RAM memory usage would keep climbing
- disk space would start going down as the amount of data in the git folder? or pack files? were growing very large
which made doing any dev work almost impossible.

---

**What worked for me regarding the Development Build** (so far?)

I followed @mjBayati [comment](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1382369333) to try to limit the mdx files needed for the development builds by using the [`ignore`](https://www.gatsbyjs.com/plugins/gatsby-source-filesystem/#ignore) option for the `gatsby-source-filesystem` plugin as [suggested here](https://paulie.dev/posts/2021/07/gatsby-slow-local-build-times/) with a [function](https://github.com/layer5io/layer5/blob/42bd0bc38856254bb0de92b9459b54baab181eff/gatsby-dev-filesystem-ignore.js) to determine which folders to ignore. You can see it in the [gatsby-config.js file](https://github.com/layer5io/layer5/blob/42bd0bc38856254bb0de92b9459b54baab181eff/gatsby-config.js), and I included an example in the dropdown.

<details>
  <summary>limit mdx files</summary>

```
 {
      resolve: "gatsby-source-filesystem",
      options: {
        path: `${__dirname}/src/collections/blog`,
        name: "blog",
        ignore: process.env.GATSBY_DEV_AMENDED === "true"
          ? devIgnoreArray("blog")
          : []
      },
    },
```

</details>

I also installed [`gatsby-plugin-no-sourcemaps` ](https://www.gatsbyjs.com/plugins/gatsby-plugin-no-sourcemaps/) based on this @pixelsoup [comment](https://github.com/gatsbyjs/gatsby/issues/36899#issuecomment-1403047762), hoping it would help.

With the above, I could navigate my development build and the disk space was stable and the amount of memory was stable at approx 3GB (which might be all the mdx files being brought into memory?).

---
last notes:
- I will try to add additional info and documentation later on, but wanted to share this initial info in case it may be helpful (and thank those who helped!). 
- I understand the dev team cannot look into this example since it is not using gatsby@next 
- Most of the above are workarounds for issues that I don't really understand, but appreciate others providing their thoughts and information.

Cheers,
Randy

**Comment by randychilau** [comment]
Hi Everyone,

The below was done based on the suggestion in this thread that the rewritten `gatsby-plugin-mdx` is bringing all mdx created pages into RAM? memory.

---

I saw some progress in making the development builds functional by reducing the file size/bundle of pages created from mdx files 

>reminder: [this project is open source](https://github.com/layer5io/layer5), and you can review any files mentioned here there.

For example, the project has an [integration template](https://github.com/layer5io/layer5/blob/master/src/templates/integrations.js) that is used by over 200 mdx files, and each of these mdx files included the same two images that had a combined file size of about 2MBs. I resized and optimized these images to less than a combined size of 200kbs. This action reduced the amount of memory used by node on my machine.

I also found a[ lightbox library](https://www.npmjs.com/package/simple-react-lightbox) being imported in some of the templates that was 300kbs large. Removing that library further reduced the memory used by node.

>A helpful [VSCode extension](https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost) to see the cost of imported 3rd party libraries is `Import Cost`.

Lastly, since the project is on React 17 and does not have access to lazy loading in React 18, I used [`loadable-components`](https://loadable-components.com/) library to lazy load below the fold components being used in templates. Note: I didn't check if this also decreased the memory used by node.

Maybe those who have hundred or thousands of mdx files without memory issues in their builds are able to keep the resulting page small in file size?

---

A couple of side notes
- when I set the MDX files to be created under Deferred Static Generation [DSG](https://www.gatsbyjs.com/docs/how-to/rendering-options/using-deferred-static-generation/) in the `createPage` call, the production build failed even with the gatsby-ssr.js workaround.
- there was an unusually large SVG file (2.7MBs) in one of the pages that caused the build to fail, after converting it to a webp image the build succeeded. I wonder if large SVGs in MDX files cause rendering/memory issues due to the number of DOM elements.

---

I hope this data point is helpful and others can share their own regarding this issue.

Any feedback, analysis, or suggestion is very welcomed, I have no experience in node memory issues and happy to learn.

Cheers,
Randy

**Comment by vsolanogo** [comment]
in my case it uses about 20+gb at peak  while on previous versions it used no more than 6gb: https://github.com/vsolanogo/apipka 

that huge amount of ram killed project and there're no way to figure out what caused it

**Comment by orval** [comment]
I've been running into out-of-memory issues in Gatsby since I started using it. I have a documentation tool that generates different kinds of output, and one of those is a Gatsby site.

It's possible I am doing something naive in terms of layout etc., but so far my attempts to optimize the configuration haven't worked out. I did a few things to change the webpack chunking but it seems like Gatsby might be overriding the settings.

So what I have done is auto-generate some fake documentation and used my tool to generate the site for it. I've checked that into this repository: https://github.com/orval/gatsbyrakoshtest

There is a `content` folder with ~4k MDX files. The total size of the folder is 24Mb. Running either `./node_modules/.bin/gatsby develop` or `./node_modules/.bin/gatsby build` results in "JavaScript heap out of memory" failures.

I'm hoping that this test case will help the maintainers look into this issue. Or at least perhaps point out something that I'm doing that is horribly wrong!
